genrl/deep/agents/dqn/base.py

```diff
@@ -0,0 +1,164 @@
+from copy import deepcopy
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import gym
+import numpy as np
+import torch
+from torch import optim as opt
+from torch.nn import functional as F
+
+from ....environments import VecEnv
+from ...common import (
+    PushReplayBuffer,
+    get_env_properties,
+    get_model,
+    safe_mean,
+    set_seeds,
+)
+from ..base import BaseAgent
+
+
+class DQN(BaseAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Union[gym.Env, VecEnv],
+        batch_size: int = 32,
+        gamma: float = 0.99,
+        layers: Tuple = (32, 32),
+        lr: float = 0.001,
+        replay_size: int = 100,
+        max_epsilon: float = 1.0,
+        min_epsilon: float = 0.01,
+        epsilon_decay: int = 1000,
+        **kwargs,
+    ):
+        super(DQN, self).__init__(
+            network_type,
+            env,
+            batch_size=batch_size,
+            gamma=gamma,
+            layers=layers,
+            **kwargs,
+        )
+        self.replay_size = replay_size
+        self.max_epsilon = max_epsilon
+        self.min_epsilon = min_epsilon
+        self.epsilon_decay = epsilon_decay
+        self.layers = layers
+        self.lr = lr
+
+        self.empty_logs()
+        self.create_model()
+
+    def create_model(self) -> None:
+        input_dim, action_dim, _, _ = get_env_properties(self.env, self.network_type)
+
+        self.model = get_model("v", self.network_type)(
+            input_dim, action_dim, "Qs", self.layers
+        )
+        self.target_model = deepcopy(self.model)
+
+        self.replay_buffer = PushReplayBuffer(self.replay_size)
```

Comment:
It would be good to take buffer type also as an input in `__init__`.

Reply:
Ok, makes sense